File: src/CWL/WorkflowDescription.fs
namespace ARCtrl.CWL

open ARCtrl.CWL
open DynamicObj
open Fable.Core

[<AttachMembers>]
type CWLWorkflowDescription(
    steps: ResizeArray<WorkflowStep>,
    inputs: ResizeArray<CWLInput>,
    outputs: ResizeArray<CWLOutput>,
    ?cwlVersion: string,
    ?requirements: ResizeArray<Requirement>,
    ?hints: ResizeArray<HintEntry>,
    ?metadata: DynamicObj,
    ?label: string,
    ?doc: string
) =
    inherit DynamicObj()

    let mutable _cwlVersion: string = cwlVersion |> Option.defaultValue "v1.2"
    let mutable _steps: ResizeArray<WorkflowStep> = steps
    let mutable _inputs: ResizeArray<CWLInput> = inputs
    let mutable _outputs: ResizeArray<CWLOutput> = outputs
    let mutable _requirements: ResizeArray<Requirement> option = requirements
    let mutable _hints: ResizeArray<HintEntry> option = hints
    let mutable _metadata: DynamicObj option = metadata
    let mutable _label: string option = label
    let mutable _doc: string option = doc

    member this.CWLVersion
        with get() = _cwlVersion
        and set(version) = _cwlVersion <- version

    member this.Steps
        with get() = _steps
        and set(steps) = _steps <- steps

    member this.Inputs
        with get() = _inputs
        and set(inputs) = _inputs <- inputs

    member this.Outputs
        with get() = _outputs
        and set(outputs) = _outputs <- outputs

    member this.Requirements
        with get() = _requirements
        and set(requirements) = _requirements <- requirements

    member this.Hints
        with get() = _hints
        and set(hints) = _hints <- hints

    member this.Metadata
        with get() = _metadata
        and set(metadata) = _metadata <- metadata

    member this.Label
        with get() = _label
        and set(label) = _label <- label

    member this.Doc
        with get() = _doc
        and set(doc) = _doc <- doc

    /// Returns workflow inputs.
    static member getInputs (workflow: CWLWorkflowDescription) =
        workflow.Inputs

    /// Returns workflow outputs.
    static member getOutputs (workflow: CWLWorkflowDescription) =
        workflow.Outputs

    /// Returns workflow requirements or an empty ResizeArray if None.
    static member getRequirementsOrEmpty (workflow: CWLWorkflowDescription) =
        workflow.Requirements |> Option.defaultValue (ResizeArray())

    /// Returns workflow hints or an empty ResizeArray if None.
    static member getHintsOrEmpty (workflow: CWLWorkflowDescription) =
        workflow.Hints |> Option.defaultValue (ResizeArray())

    /// Returns the workflow's hints, creating and assigning a new empty ResizeArray if None.
    static member getOrCreateHints (workflow: CWLWorkflowDescription) =
        match workflow.Hints with
        | Some hints -> hints
        | None ->
            let hints = ResizeArray()
            workflow.Hints <- Some hints
            hints

File: js/CWL/WorkflowDescription.js
import { defaultArg, unwrap } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { DynamicObj_$reflection, DynamicObj } from "../fable_modules/DynamicObj.4.0.3/DynamicObj.fs.js";
import { class_type } from "../fable_modules/fable-library-js.4.22.0/Reflection.js";

export class CWLWorkflowDescription extends DynamicObj {
    constructor(steps, inputs, outputs, cwlVersion, requirements, hints, metadata) {
        super();
        this._cwlVersion = defaultArg(cwlVersion, "v1.2");
        this._steps = steps;
        this._inputs = inputs;
        this._outputs = outputs;
        this._requirements = requirements;
        this._hints = hints;
        this._metadata = metadata;
    }
    get CWLVersion() {
        const this$ = this;
        return this$._cwlVersion;
    }
    set CWLVersion(version) {
        const this$ = this;
        this$._cwlVersion = version;
    }
    get Steps() {
        const this$ = this;
        return this$._steps;
    }
    set Steps(steps) {
        const this$ = this;
        this$._steps = steps;
    }
    get Inputs() {
        const this$ = this;
        return this$._inputs;
    }
    set Inputs(inputs) {
        const this$ = this;
        this$._inputs = inputs;
    }
    get Outputs() {
        const this$ = this;
        return this$._outputs;
    }
    set Outputs(outputs) {
        const this$ = this;
        this$._outputs = outputs;
    }
    get Requirements() {
        const this$ = this;
        return unwrap(this$._requirements);
    }
    set Requirements(requirements) {
        const this$ = this;
        this$._requirements = requirements;
    }
    get Hints() {
        const this$ = this;
        return unwrap(this$._hints);
    }
    set Hints(hints) {
        const this$ = this;
        this$._hints = hints;
    }
    get Metadata() {
        const this$ = this;
        return unwrap(this$._metadata);
    }
    set Metadata(metadata) {
        const this$ = this;
        this$._metadata = metadata;
    }
}

export function CWLWorkflowDescription_$reflection() {
    return class_type("ARCtrl.CWL.CWLWorkflowDescription", undefined, CWLWorkflowDescription, DynamicObj_$reflection());
}

export function CWLWorkflowDescription_$ctor_704BBD06(steps, inputs, outputs, cwlVersion, requirements, hints, metadata) {
    return new CWLWorkflowDescription(steps, inputs, outputs, cwlVersion, requirements, hints, metadata);
}

